### ISSUE TITLE
New beam-beam user interface

### ISSUE THREAD

**Issue by rdemaria** [SixTrack/SixTrack#11]
Beam-beam lens parameter will be always read from input and never inferred.

**Comment by kyrsjo** [comment]
This is in the process of being merged in this branch:
https://github.com/kyrsjo/SixTrack/tree/JavierBB

**Comment by kyrsjo** [comment]
Todo: Check/add CRLIBM-compatible input file parsing (conversion of string->float)

**Comment by kyrsjo** [comment]
Code review strangeness: Why is ed and ek getting modified outside of reading the single element list / daten?

Se "note" in https://github.com/kyrsjo/SixTrack/commit/b07ff5bfe8a754d25d0b22fe7fefac83a05f881c

**Comment by kyrsjo** [comment]
Useless variables:
* bbbx,bbby,bbbs from fort.2?
* ibb6d
* Eventually: lhc

**Comment by kyrsjo** [comment]
To check: What happens if parbe(ix,2).le.0, i.e. negative number of slices? See also change at end of block UMLALID1:
```
          if(parbe(ix,2).ge.0d0) then !With or without slicing
            do ii=1,10
              bbcu(ibb,ii)=bbcu(ibb,ii)*c1m6
            enddo
          endif
```
Used to be (i.e. with .gt. not .ge.):
```
          if(parbe(ix,2).gt.0d0) then                                    !hr08
            do ii=1,10
              bbcu(ibb,ii)=bbcu(ibb,ii)*c1m6
            enddo
          endif
```

All other places in the code it is .gt.?

**Comment by kyrsjo** [comment]
Relevant MadX tracker entry: https://svnweb.cern.ch/trac/madx/ticket/380#no1

**Comment by kyrsjo** [comment]
Good news! After printing the contents of the BBCU array when it is first set (in the ULALID1 block) in the old version, and putting this in as the corresponding mm1..mm10 elements in fort.3, I get the same closed orbit and tracking results, even with extra detailed debugging printout.

The fort.3 from HO_6D WITH the new BB block, changed to match the version with the old BB block:
[fort.3.txt](https://github.com/SixTrack/SixTrack/files/53901/fort.3.txt)

The only difference left is in the postprocessing, where the "nonlinear detuning" in X in the first turn is now 0.94354-310, while in the old version it was 0.20085D+01. Will investigate...

**Comment by kyrsjo** [comment]
OK, that was fixed by increasing the buffer size sent to splitfld from subroutine SUMPOS! Fort.10 was the same in both cases, but the numbers got mangled in splitfld/fround.

Furthermore, I stripped off the extra "0"s at the end of the numbers first line of the BEAM block entries in fort.3, and it worked with the "standard" JavierBB executable!

Test cases:
* The current BB block: 
[HO_6D_orig_fort.2.txt](https://github.com/SixTrack/SixTrack/files/54284/HO_6D_orig_fort.2.txt)
[HO_6D_orig_fort.3.txt](https://github.com/SixTrack/SixTrack/files/54282/HO_6D_orig_fort.3.txt)
[HO_6D_orig_fort.34.txt](https://github.com/SixTrack/SixTrack/files/54283/HO_6D_orig_fort.34.txt)
* New BB block:
[HO_6D_JBB_fort.2.txt](https://github.com/SixTrack/SixTrack/files/54285/HO_6D_JBB_fort.2.txt)
[HO_6D_JBB_fort.3.txt](https://github.com/SixTrack/SixTrack/files/54287/HO_6D_JBB_fort.3.txt)
[HO_6D_JBB_fort.34.txt](https://github.com/SixTrack/SixTrack/files/54286/HO_6D_JBB_fort.34.txt)

**Comment by kyrsjo** [comment]
HO_4D test now working; fixing it required (1) updating fort.3 and (2) making two changes to sixtrack.s:
* In block "umlalid1", replace ```if(parbe(ix,2).ge.0d0)``` with ```if(parbe(ix,2).gt.0d0)```; If the intention was to accept 6D/0 slices (is that even possible?) as a 6D case and not a 4D case, rather use parbe(:,17).
* In blocks "beamco" and "beamcoo", change ed(ix) and ek(ix) to parbe(ix,5) and parbe(ix,6); otherwise crkveb and cikveb would not be set correctly and tracking would return incorrect results

Test cases:
* The current BB block: 
[HO_4D_orig_fort.2.txt](https://github.com/SixTrack/SixTrack/files/56742/HO_4D_orig_fort.2.txt)
[HO_4D_orig_fort.3.txt](https://github.com/SixTrack/SixTrack/files/56743/HO_4D_orig_fort.3.txt)
[HO_4D_orig_fort.34.txt](https://github.com/SixTrack/SixTrack/files/56744/HO_4D_orig_fort.34.txt)
* New BB block: 
[HO_4D_JBB_fort.2.txt](https://github.com/SixTrack/SixTrack/files/56755/HO_4D_JBB_fort.2.txt)
[HO_4D_JBB_fort.3.txt](https://github.com/SixTrack/SixTrack/files/56756/HO_4D_JBB_fort.3.txt)
[HO_4D_JBB_fort.34.txt](https://github.com/SixTrack/SixTrack/files/56754/HO_4D_JBB_fort.34.txt)

**Comment by kyrsjo** [comment]
Standard (not debug) version now also passes HO_4D test:
https://github.com/kyrsjo/SixTrack/commit/9f726530e82f59447f4123a246b9daea26d85799

**Comment by rdemaria** [comment]
For 4D/6D the intentions were:  0 means 4D, 1 means 6D one slice, <n> means 6D <n> slices

**Comment by kyrsjo** [comment]
Great, that is what I thought as well (and what the updated manual says). I just remembered there was some strange stuff going on with 4D/6D and slicing, but it was probably just to allow 1 slice being used for 6D?

If so, then my fix is correct, but it doesn't explain why the bug was introduced in the first place, or why the unit conversion is needed for 6D but not 4D. A comment line would be useful.

Now working on fixing fort.3 for the HO_LR test. It has a lot of entries. Once that is done an it passes, I will try to remove useless variables, possibly including the last 3 columns read from fort.3's single element list (this might also require a MadX change, to avoid writing them out, unless we teach SixTrack to just ignore these 3 columns if they are there)...

**Comment by kyrsjo** [comment]
## Poor-mans documentation: How to convert 4D and 6D BB from old BEAM block/fort.2 to new BEAM block

* Fort.2: All columns except the name and type should be set to 0.0

### 4D Beam Beam element:
1. In an "old" version of SixTrack, in the block ```UMLALID1``` write out BBCU(ibb,:) just after initialization, i.e. insert the lines

>              write(*,"(1x,a4,i7,a3,1x,12g30.20)") "bbcu(",ibb,",:)",
>     &            (bbcu(ibb,j),j=1,12)

just after the block

>           if(lhc.eq.1) then
>            dummy=bbcu(ibb,1)
>            bbcu(ibb,1)=bbcu(ibb,2)
>            bbcu(ibb,2)=dummy
>            dummy=bbcu(ibb,4)
>            bbcu(ibb,4)=bbcu(ibb,9)
>            bbcu(ibb,9)=dummy
>            dummy=bbcu(ibb,5)
>            bbcu(ibb,5)=bbcu(ibb,7)
>            bbcu(ibb,7)=dummy
>            dummy=bbcu(ibb,6)
>            bbcu(ibb,6)=bbcu(ibb,10)
>            bbcu(ibb,10)=dummy
>          endif
>          if(lhc.eq.2) then
>            bbcu(ibb,1)=bbbx(ix)
>            bbcu(ibb,2)=bbby(ix)
>            bbcu(ibb,3)=bbbs(ix)
>          endif 

2. Copy the first two (and only non-zero) elements of bbcu into fort.3 for the element (column # 3 and # 4, first floating point entries)
3. The two last columns for the element in fort.3 can be found as ed and ek from fort.2 (these should be then set to 0.0 in fort.2).

### 6D Beam-beam element:
1. Insert the same block to printout bbcu
2. The sigmas (line 2 and 3 of the element in fort.3) are given by bbcu with indices:
 * mm1 = bbcu(ibb,1)
 * mm2 = bbcu(ibb,4)
 * mm3 = bbcu(ibb,6)
 * mm4 = bbcu(ibb,2)
 * mm5 = bbcu(ibb,9)
 * mm6 = bbcu(ibb,10)
 * mm7 = bbcu(ibb,3)
 * mm8 = bbcu(ibb,5)
 * mm9 = bbcu(ibb,7)
 * mm10 = bbcu(ibb,8)
3. Copy over the data from fort.2 into the first line of the new BB element in fort.3

**Comment by kyrsjo** [comment]
The HO_LR test is now also passing. This means all tests are now passing with the modified code!

* Old BB block version:
[HO_LR_orig_fort.2.txt](https://github.com/SixTrack/SixTrack/files/58077/HO_LR_orig_fort.2.txt)
[HO_LR_orig_fort.3.txt](https://github.com/SixTrack/SixTrack/files/58075/HO_LR_orig_fort.3.txt)
[HO_LR_orig_fort.34.txt](https://github.com/SixTrack/SixTrack/files/58076/HO_LR_orig_fort.34.txt)
* New BB block version:
[HO_LR_JBB_fort.2.txt](https://github.com/SixTrack/SixTrack/files/58078/HO_LR_JBB_fort.2.txt)
[HO_LR_JBB_fort.3.txt](https://github.com/SixTrack/SixTrack/files/58080/HO_LR_JBB_fort.3.txt)
[HO_LR_JBB_fort.34.txt](https://github.com/SixTrack/SixTrack/files/58079/HO_LR_JBB_fort.34.txt)

**Comment by kyrsjo** [comment]
Semi-automatic way of converting the BEAM block:
1. Checkout this version of the code & run it with an OLD beam block ```fort.3```: https://github.com/SixTrack/SixTrack/commit/49f3bfe728fd5ec0da7a39a0082c713f7936b2e0
2. Compile and run it, pipe standard output to a new file "log.txt"
3. In the output, look for lines like

>   NEW BEAM BEAM DATA 
> bb_par.l1b1_1    0     0.31146284209145926047E-01    0.31280226459102507264E-01   
> 0.21054413559999998751E-04     2.2066171059999999393         1.0000000000000000000  

and 

>   NEW BEAM BEAM DATA 
> bb_ho1b1_0         15     0.29500000000000001481E-03     1.5700000000000000622        0.96983435799999992302E-06   -0.13760678739999999771E-05
>    0.50520080285947416223E-04   -0.60102042820216678839E-06    0.22227616225043104331E-02    0.50286882391145739421E-04    0.77036976500561834648E-06
>    0.22330904282531743171E-02   -0.36528061007530750743E-08    0.26610093044576254908E-07    0.36386713069639949989E-07    0.11594620089159194493E-06     1.0000000000000000000  

Copy ALL those lines (remember the last one into the BEAM block of your new fort.3, replacing everything but the FIRST line of the old BEAM block. Clean up by removing the "NEW BEAM BEAM DATA" lines (these are just there to make it easier to find in log.out)
4. Run the simulation with a version of SixTrack supporting the new BEAM block. The closed orbit and tracking should give the same results.

This procedure was tested on HO_LR, which contains both 4D and 6D elements.

**Comment by kyrsjo** [comment]
All SixTest tests successfully converted, with the exception of "Javier", which is not compatible with the new input block. Also converted but not checked in is bnl, where the file "beambeam-output.dat" is slightly different from the canonical but identical to the one produced by the "master" branch.

**Comment by rdemaria** [comment]
@kyrsjo  what modifying 49f3bfe and make it produce a file called fort.3.beambeam collecting all the lines needed?

**Comment by kyrsjo** [comment]
Hi Riccardo;

Unless there is something I don't understand about bnlelens, I think I have the input working fine: If I run the old version (branch "master") and the new version (branch "JavierBB") with unconverted/converted output, I get the same results. However, neither of them are the same as the canonical.

So it is pretty likely that the problem is not in JavierBB, but rather that the main SixTrack changed somewhere along the way (or the compiler did...). Unfortunately, it seems the bnl test was never run (at least not by me), since trying to run it with a non-"bnlelens" executable just produces the error message "Ignoring TEST bnl as this executable does not handle bnlelens." instead of SUCCESS, which is pretty easy to miss...

**Comment by kyrsjo** [comment]
99% done and mostly ready to do the merge (once MadX is ready @ldeniau ): Once the "javier" and "bnl" tests are decided on, we can bump the version number and merge. We may also want to add the HO_6D/4D/LR tests provided by Javier to the SixTests directory.

I removed the last 3 columns from the single element list (if they are present, they are just ignored) and also removed the ibb6d flag. They should be removed from the MadX output.

All tests (except the two mentioned) are now passing.

**Comment by kyrsjo** [comment]
BNL test mostly-working converted inputs:
[bnl_fort.2.txt](https://github.com/SixTrack/SixTrack/files/63848/bnl_fort.2.txt)
[bnl_fort.3.txt](https://github.com/SixTrack/SixTrack/files/63849/bnl_fort.3.txt)

**Comment by rdemaria** [comment]
Decision:
  * restore backward compatibility
  * old block also write input for new block 
  * introduce "expert" keyword after BEAM for new input

**Comment by kyrsjo** [comment]
Now being fixed in #246

**Comment by kyrsjo** [comment]
Fixed in #246
